### ISSUE TITLE
Support passing REPLICA IDENTITY FULL old-remote-row info to conflict triggers

### ISSUE THREAD

**Issue by ringerc** [2ndQuadrant/bdr#104]
Presently we can't really do change merging in conflict triggers because we only have the new old row; we don't know exactly what changed and from what old values.

If `REPLICA IDENTITY FULL` information exists on the remote end we should use that, sending it to the conflict trigger as the old remote row, so the conflict trigger can see what's changed and do intelligent merging. This would enable things like merging a remote `col = col + 1` and local `col = col + 3` into `col = col + 4`

**Comment by anarazel** [comment]
> If `REPLICA IDENTITY FULL` information exists on the remote end we should use that, sending it to the conflict trigger as the old remote row, so the conflict trigger can see what's changed and do intelligent merging. This would enable things like merging a remote `col = col + 1` and local `col = col + 3` into `col = col + 4`

It does not. You cannot discern col = 15334 and col = col + 1 on that
basis alone.

**Comment by ringerc** [comment]
​So REPLICA IDENTITY FULL is not sufficient to get both "OLD" and "NEW"
states for both local and remote tuples?

**Comment by anarazel** [comment]
> ​So REPLICA IDENTITY FULL is not sufficient to get both "OLD" and "NEW"
> states for both local and remote tuples?

It is, you just can't infer the difference between SET foo = foo + 3 and
SET foo = 343434 by just looking at the tuples.

**Comment by ringerc** [comment]
On 3 July 2015 at 15:19, Andres Freund <notifications@github.com> wrote:

> > ​So REPLICA IDENTITY FULL is not sufficient to get both "OLD" and "NEW"
> > states for both local and remote tuples?
>
> It is, you just can't infer the difference between SET foo = foo + 3 and
> SET foo = 343434 by just looking at the tuples.
>
>
In the general case I absolutely agree.

If you have application-specific knowledge that the app only ever
increments the field, though, you could then merge local and remote
changes. It'd be interesting for monotonically increasing fields, that sort
of thing.

 It's more interesting for cases where the remote row changes field X, the
local row changes field Y, and your conflict trigger knows that fields X
and Y are independent.

I don't think BDR its self could ever safely or sanely merge rows, but I'm
inclined to think there are uses for apps. So it'd be nice to expose the
OLD remote row to conflict triggers where REPLICA IDENTITY FULL permits
that.

This is mainly a case of "I'll note that down while I have it in mind"
rather than something I'll tackle soon, but I'll try to add room for OLD
remote rows when I next touch the conflict trigger interfaces.


-- 
 Craig Ringer                   http://www.2ndQuadrant.com/
 PostgreSQL Development, 24x7 Support, Training & Services

**Comment by sym39** [comment]
Should I understand that using REPLICA IDENTITY FULL is not supported for now?
I just tried to use it, it does not work at all for me, I get the following message:

    mydb=# create table test (i int primary key, t text);
    CREATE TABLE
    mydb=# insert into test values (1, 'hello');
    INSERT 0 1
    mydb=# update test set t='test' where i=1;
    UPDATE 1
    mydb=# ALTER TABLE test REPLICA IDENTITY FULL;
    ALTER TABLE
    mydb=# update test set t='test' where i=1;
    ERROR:  Cannot run UPDATE or DELETE on table test because it does not have primary key.
    HINT:  Add primary key to the table
    mydb=# ALTER TABLE test REPLICA IDENTITY DEFAULT;
    ALTER TABLE
    mydb=# update test set t='test' where i=1;
    UPDATE 1

Is is the same issue reported here? I am using 0.9.1 version.

#83 gives me the hint to the logical decoding feature to be notified on tuple change, but it would be very nice if IDENTITY FULL would work to emulate what would do a trigger (to fire only if a particular field has changed).

**Comment by anarazel** [comment]
> Should I understand that using REPLICA IDENTITY FULL is not supported
> for now?

Correct.

> #83 gives me the hint to the logical decoding feature to be notified
> on tuple change, but it would be very nice if IDENTITY FULL would work
> to emulate what would do a trigger (to fire only if a particular field
> has changed).

IDENTITY FULL won't make it easier in any way to trigger things on the
remote side.

**Comment by sym39** [comment]
> IDENTITY FULL won't make it easier in any way to trigger things on the remote side.
Sorry for the 'trigger shortcut', I don't want to reimplement all trigger things in the remote, simply, I think It would have help in my specific use case (only one writer at a time) to fire an event because one specific somewhat 'important' field of a tuple has been modified.

**Comment by ringerc** [comment]
@sym39 I think you're confusing conflict triggers for normal PostgreSQL triggers.

A conflict "trigger" is a function that BDR invokes when it detects that a write from a remote node conflicts with a concurrent write locally or from another remote node.

It's not fired for every write received from a remote node, and it's not a general purpose trigger that can go and do whatever it wants. These functions are meant to override the default conflict handling in BDR with application-specific logic, that's all.

While `REPLICA IDENTITY FULL` would be nice to have for #83, it's a different matter to what's being discussed here.

It seems the error message when `REPLICA IDENTITY FULL` is enabled could use some improvement. I'll take a look once I've got 0.9.2 sorted.

**Comment by sym39** [comment]
>  @sym39 I think you're confusing conflict triggers for normal PostgreSQL triggers.

You are right. 
What interests me is to find a way to be notified on changes, like what is described on #83. My goal is to use the test_decoding extension to get the changes (old tuples / new tuples) and refresh a cache accordingly. Not having the old values is a bit problematic because not all the fields are updated at the same time, and I would like to avoid unnecessary cache refreshs.

Should I open a dedicated ticket for the REPLICA IDENTITY FULL problem described above? (The strange error about the primary key not present)?

**Comment by ringerc** [comment]
@sym39 Feel free, that'd be useful

**Comment by anarazel** [comment]
> What interests me is to find a way to be notified on changes, like
> what is described on #83. My goal is to use the test_decoding
> extension to get the changes (old tuples / new tuples) and refresh a
> cache accordingly. Not having the old values is a bit problematic
> because not all the fields are updated at the same time, and I would
> like to avoid unnecessary cache refreshs.

That sounds pretty much entirely unrelated to bdr?

**Comment by ringerc** [comment]
Yep, I think @sym39 probably wants one of the logical decoding to json extensions under development in the community.

**Comment by sym39** [comment]
Sorry to tell all about my life ;-)
But I think the error I get when using the REPLICA IDENTITY FULL is related to BDR, isn't it?

    mydb=# ALTER TABLE test REPLICA IDENTITY FULL;
    ALTER TABLE
    mydb=# update test set t='test' where i=1;
    ERROR:  Cannot run UPDATE or DELETE on table test because it does not have primary key.
    HINT:  Add primary key to the table
    mydb=# ALTER TABLE test REPLICA IDENTITY DEFAULT;
    ALTER TABLE
    mydb=# update test set t='test' where i=1;
    UPDATE 1

**Comment by ringerc** [comment]
Yes, the error should more clearly identify the actual issue; feel free to open a bug.

**Comment by sym39** [comment]
> Yes, the error should more clearly identify the actual issue; feel free to open a bug.

Done => #105
